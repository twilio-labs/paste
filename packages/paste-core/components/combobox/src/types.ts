import {
  UseComboboxPrimitiveProps,
  UseComboboxPrimitiveState,
  UseComboboxPrimitiveReturnValue,
} from '@twilio-paste/combobox-primitive';
import {InputVariants, InputProps} from '@twilio-paste/input';

type Item = string | {[key: string]: any};

export interface ComboboxProps extends Omit<InputProps, 'id' | 'type' | 'value'> {
  autocomplete?: boolean;
  helpText?: string | React.ReactNode;
  initialIsOpen?: UseComboboxPrimitiveProps<Item>['initialIsOpen'];
  initialSelectedItem?: UseComboboxPrimitiveProps<Item>['initialSelectedItem'];
  items: UseComboboxPrimitiveProps<Item>['items'];
  itemToString?: UseComboboxPrimitiveProps<Item>['itemToString'];
  labelText: string | NonNullable<React.ReactNode>;
  onHighlightedIndexChange?: UseComboboxPrimitiveProps<Item>['onHighlightedIndexChange'];
  onInputValueChange?: UseComboboxPrimitiveProps<Item>['onInputValueChange'];
  onIsOpenChange?: UseComboboxPrimitiveProps<Item>['onIsOpenChange'];
  onSelectedItemChange?: UseComboboxPrimitiveProps<Item>['onSelectedItemChange'];
  optionTemplate?: (item: string | {}) => React.ReactNode;
  groupLabelTemplate?: (groupName: string) => React.ReactNode;
  selectedItem?: UseComboboxPrimitiveProps<Item>['selectedItem'];
  inputValue?: UseComboboxPrimitiveProps<Item>['inputValue'];
  groupItemsBy?: string;
  variant?: InputVariants;
  state?: Partial<UseComboboxPrimitiveReturnValue<Item>>;
}

export interface RenderItemProps extends Pick<ComboboxProps, 'optionTemplate'> {
  item: Item;
  index: number;
  getItemProps: any;
  highlightedIndex: UseComboboxPrimitiveState<Item>['highlightedIndex'];
  inGroup?: boolean;
  optionUID: (id: any) => string;
}

export interface RenderItemsProps extends Omit<RenderItemProps, 'item' | 'index'> {
  items: Item[];
}

export interface RenderGroupItemsProps
  extends RenderItemsProps,
    Pick<ComboboxProps, 'groupLabelTemplate' | 'groupItemsBy'> {
  groupUID: (id: any) => string;
}

export type RenderListBoxProps = RenderGroupItemsProps;
