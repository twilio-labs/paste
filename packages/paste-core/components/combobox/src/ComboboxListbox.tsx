import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';

export interface ComboboxListboxProps {
  children: NonNullable<React.ReactNode>;
}

const ComboboxListbox = React.forwardRef<HTMLUListElement, ComboboxListboxProps>(({children, ...props}, ref) => {
  // TODO:
  // Right now the listbox is absolutely positioned. This will
  // fail if a combobox is placed at the bottom of a screen
  // so we need upgrade the positioning using popperjs in the future.
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      as="ul"
      backgroundColor="colorBackgroundBody"
      borderBottomLeftRadius="borderRadius20"
      borderBottomRightRadius="borderRadius20"
      boxShadow="shadowCard"
      listStyleType="none"
      margin="space0"
      overflow="hidden"
      padding="space0"
      position="absolute"
      width="100%"
      zIndex="zIndex10"
      marginTop="space20"
      ref={ref}
    >
      {children}
    </Box>
  );
});

ComboboxListbox.displayName = 'ComboboxListbox';

if (process.env.NODE_ENV === 'development') {
  ComboboxListbox.propTypes = {
    children: PropTypes.node.isRequired,
  };
}

export {ComboboxListbox};
