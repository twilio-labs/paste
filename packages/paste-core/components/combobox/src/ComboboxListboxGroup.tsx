import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import {Text} from '@twilio-paste/text';
import {ComboboxProps} from './types';

export interface ComboboxListboxGroupProps extends Pick<ComboboxProps, 'groupLabelTemplate'> {
  children: NonNullable<React.ReactNode>;
  groupName: string;
}

const ComboboxListboxGroup = React.forwardRef<HTMLDivElement, ComboboxListboxGroupProps>(
  ({children, groupName, groupLabelTemplate}, ref) => {
    return (
      <Box as="div" role="group" aria-label={groupName} ref={ref}>
        <Box as="div" role="presentation" paddingY="space30" paddingX="space70">
          <Text as="span" fontWeight="fontWeightBold">
            {groupLabelTemplate ? groupLabelTemplate(groupName) : groupName}
          </Text>
        </Box>
        {children}
      </Box>
    );
  }
);

ComboboxListboxGroup.displayName = 'ComboboxListboxGroup';

ComboboxListboxGroup.propTypes = {
  children: PropTypes.node.isRequired,
  groupName: PropTypes.string.isRequired,
};

export {ComboboxListboxGroup};
