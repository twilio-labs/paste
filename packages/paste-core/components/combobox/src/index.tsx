export {useComboboxPrimitive as useCombobox} from '@twilio-paste/combobox-primitive';
export type {UseComboboxPrimitiveState as UseComboboxState} from '@twilio-paste/combobox-primitive';
export * from './Combobox';
export * from './styles/ComboboxInputWrapper';
export * from './styles/ComboboxListbox';
export * from './styles/ComboboxListboxOption';
export * from './styles/ComboboxListboxGroup';
export type {ComboboxProps} from './types';
