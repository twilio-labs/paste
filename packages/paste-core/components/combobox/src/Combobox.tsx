import * as React from 'react';
import * as PropTypes from 'prop-types';
import {css, styled} from '@twilio-paste/styling-library';
import {useUID, useUIDSeed} from 'react-uid';
import {useComboboxPrimitive} from '@twilio-paste/combobox-primitive';
import {ChevronDownIcon} from '@twilio-paste/icons/esm/ChevronDownIcon';
import {Box} from '@twilio-paste/box';
import {InputBox, InputChevronWrapper} from '@twilio-paste/input-box';
import {Label} from '@twilio-paste/label';
import {HelpText, HelpTextVariants} from '@twilio-paste/help-text';
import {InputElement, InputVariants} from '@twilio-paste/input';
import {TextColor} from '@twilio-paste/style-props';
import {ComboboxInputWrapper} from './ComboboxInputWrapper';
import {ComboboxListbox} from './ComboboxListbox';
import {ComboboxListboxGroup} from './ComboboxListboxGroup';
import {ComboboxListboxOption} from './ComboboxListboxOption';
import {RenderGroupItemsProps, RenderItemProps, RenderItemsProps, RenderListBoxProps, ComboboxProps} from './types';

// This module can only be referenced with ECMAScript imports/exports by turning on the 'esModuleInterop' flag and referencing its default export
const groupBy = require('lodash.groupby');

// Fixes chevron overlapping really long text
/* eslint-disable emotion/syntax-preference */
const StyledInputAsSelect = styled(InputElement)(
  css({
    paddingRight: 'space100',
  })
);
/* eslint-enable */

const renderItem = ({
  item,
  index,
  getItemProps,
  highlightedIndex,
  optionTemplate,
  inGroup,
  optionUID,
}: RenderItemProps): React.ReactNode => {
  return (
    <ComboboxListboxOption
      {...getItemProps({item, index})}
      highlighted={highlightedIndex === index}
      key={optionUID(index)}
      variant={inGroup ? 'groupOption' : 'default'}
    >
      {optionTemplate ? optionTemplate(item) : item}
    </ComboboxListboxOption>
  );
};

const renderItems = ({
  items,
  getItemProps,
  highlightedIndex,
  optionTemplate,
  optionUID,
}: RenderItemsProps): React.ReactNode[] =>
  items.map((item, index) => {
    return renderItem({item, index, getItemProps, highlightedIndex, optionTemplate, optionUID});
  });

const renderGroupedItems = ({
  items,
  getItemProps,
  highlightedIndex,
  optionTemplate,
  groupLabelTemplate,
  groupItemsBy,
  optionUID,
  groupUID,
}: RenderGroupItemsProps): React.ReactNode[] | null => {
  if (groupItemsBy != null) {
    const groupedItems = groupBy(items, (item: {}) => item[groupItemsBy]);
    let itemIndex = 0;
    return Object.keys(groupedItems).map(group => {
      if (group === 'undefined') {
        return groupedItems[group].map((item: {}) => {
          return renderItem({item, index: itemIndex++, getItemProps, highlightedIndex, optionTemplate, optionUID});
        });
      }
      return (
        <ComboboxListboxGroup groupName={group} groupLabelTemplate={groupLabelTemplate} key={groupUID(group)}>
          {groupedItems[group].map((item: {}) => {
            return renderItem({
              item,
              index: itemIndex++,
              getItemProps,
              highlightedIndex,
              optionTemplate,
              inGroup: true,
              optionUID,
            });
          })}
        </ComboboxListboxGroup>
      );
    });
  }
  return null;
};

const renderListBox = ({
  items,
  getItemProps,
  highlightedIndex,
  optionTemplate,
  groupLabelTemplate,
  groupItemsBy,
  optionUID,
  groupUID,
}: RenderListBoxProps): React.ReactNode[] | null =>
  groupItemsBy
    ? renderGroupedItems({
        items,
        getItemProps,
        highlightedIndex,
        optionTemplate,
        groupItemsBy,
        groupLabelTemplate,
        optionUID,
        groupUID,
      })
    : renderItems({items, getItemProps, highlightedIndex, optionTemplate, optionUID});

const getHelpTextVariant = (variant: InputVariants, hasError: boolean | undefined): HelpTextVariants => {
  if (hasError && variant === 'inverse') {
    return 'error_inverse';
  }
  if (hasError) {
    return 'error';
  }
  if (variant === 'inverse') {
    return 'inverse';
  }
  return 'default';
};

const Combobox = React.forwardRef<HTMLInputElement, ComboboxProps>(
  (
    {
      autocomplete,
      disabled,
      hasError,
      helpText,
      initialSelectedItem,
      initialIsOpen,
      inputValue,
      insertAfter,
      insertBefore,
      items,
      itemToString,
      labelText,
      onInputValueChange,
      onSelectedItemChange,
      optionTemplate,
      onHighlightedIndexChange,
      onIsOpenChange,
      required,
      selectedItem,
      groupItemsBy,
      groupLabelTemplate,
      variant = 'default',
      state,
      ...props
    },
    ref
  ) => {
    const {
      getComboboxProps,
      getInputProps,
      getItemProps,
      getLabelProps,
      getMenuProps,
      getToggleButtonProps,
      highlightedIndex,
      isOpen,
    } =
      state ||
      useComboboxPrimitive({
        initialSelectedItem,
        items,
        onHighlightedIndexChange,
        onInputValueChange,
        onIsOpenChange,
        onSelectedItemChange,
        ...(itemToString && {itemToString}),
        ...(initialIsOpen && {initialIsOpen}),
        ...(inputValue && {inputValue}),
        ...(selectedItem && {selectedItem}),
      });

    if (
      getComboboxProps === undefined ||
      getInputProps === undefined ||
      getItemProps === undefined ||
      getLabelProps === undefined ||
      getMenuProps === undefined ||
      getToggleButtonProps === undefined ||
      highlightedIndex === undefined ||
      isOpen === undefined
    ) {
      throw new Error(
        '[Combobox]: One of getComboboxProps, getInputProps, getItemProps, getLabelProps, getMenuProps, getToggleButtonProps, highlightedIndex or isOpen is missing from the state object. Please make sure this is provided.'
      );
    }

    const helpTextId = useUID();
    const groupUID = useUIDSeed();
    const optionUID = useUIDSeed();

    let iconColor = 'colorTextIcon' as TextColor;
    if (disabled) {
      iconColor = 'colorTextWeaker';
    } else if (variant === 'inverse') {
      iconColor = 'colorTextInverseWeak';
    }

    return (
      <Box position="relative">
        <Label disabled={disabled} required={required} variant={variant} {...getLabelProps()}>
          {labelText}
        </Label>
        <InputBox
          disabled={disabled}
          hasError={hasError}
          insertBefore={insertBefore}
          insertAfter={insertAfter}
          variant={variant}
        >
          <ComboboxInputWrapper {...getComboboxProps({role: 'combobox'})}>
            <StyledInputAsSelect
              {...getToggleButtonProps({tabIndex: 0})}
              {...getInputProps({disabled, ref})}
              {...(!autocomplete ? {onChange: event => event.preventDefault()} : undefined)}
              aria-describedby={helpTextId}
              {...props}
              type="text"
              paddingRight="space90"
            />
            <InputChevronWrapper>
              <ChevronDownIcon aria-hidden="true" decorative color={iconColor} size="sizeIcon30" />
            </InputChevronWrapper>
          </ComboboxInputWrapper>
        </InputBox>
        <ComboboxListbox hidden={!isOpen} {...getMenuProps()}>
          {renderListBox({
            items,
            getItemProps,
            highlightedIndex,
            optionTemplate,
            groupItemsBy,
            groupLabelTemplate,
            optionUID,
            groupUID,
          })}
        </ComboboxListbox>
        {helpText && (
          <HelpText id={helpTextId} variant={getHelpTextVariant(variant, hasError)}>
            {helpText}
          </HelpText>
        )}
      </Box>
    );
  }
);

Combobox.displayName = 'Combobox';

if (process.env.NODE_ENV === 'development') {
  Combobox.propTypes = {
    autocomplete: PropTypes.bool,
    groupItemsBy: PropTypes.string,
    helpText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    initialIsOpen: PropTypes.bool,
    initialSelectedItem: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    items: PropTypes.arrayOf(PropTypes.any).isRequired,
    itemToString: PropTypes.func,
    labelText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    onHighlightedIndexChange: PropTypes.func,
    onInputValueChange: PropTypes.func,
    onIsOpenChange: PropTypes.func,
    onSelectedItemChange: PropTypes.func,
    optionTemplate: PropTypes.func,
    selectedItem: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  };
}

export {Combobox};
