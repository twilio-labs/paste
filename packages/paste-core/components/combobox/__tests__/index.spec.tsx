import * as React from 'react';
import _ from 'lodash';
import {axe} from 'jest-axe';
import {render, screen, fireEvent} from '@testing-library/react';
import {Button} from '@twilio-paste/button';
import {CloseIcon} from '@twilio-paste/icons/esm/CloseIcon';
import {Box} from '@twilio-paste/box';
import {useCombobox, Combobox} from '../src';
import {ComboboxProps} from '../src/types';

const items = ['Alert', 'Anchor', 'Button', 'Card', 'Heading', 'List', 'Modal', 'Paragraph'];

const objectItems = [
  {code: 'AD', label: 'Andorra', phone: '376'},
  {code: 'AE', label: 'United Arab Emirates', phone: '971'},
  {code: 'AF', label: 'Afghanistan', phone: '93'},
  {code: 'AG', label: 'Antigua and Barbuda', phone: '1-268'},
  {code: 'AI', label: 'Anguilla', phone: '1-264'},
  {code: 'AL', label: 'Albania', phone: '355'},
  {code: 'AM', label: 'Armenia', phone: '374'},
  {code: 'AO', label: 'Angola', phone: '244'},
  {code: 'AQ', label: 'Antarctica', phone: '672'},
  {code: 'AR', label: 'Argentina', phone: '54'},
  {code: 'AS', label: 'American Samoa', phone: '1-684'},
  {code: 'AT', label: 'Austria', phone: '43'},
];

const groupedItems = [
  {group: 'Components', label: 'Alert'},
  {group: 'Components', label: 'Anchor'},
  {group: 'Components', label: 'Button'},
  {group: 'Components', label: 'Card'},
  {group: 'Components', label: 'Heading'},
  {group: 'Components', label: 'List'},
  {group: 'Components', label: 'Modal'},
  {group: 'Components', label: 'Paragraph'},
  {group: 'Primitives', label: 'Box'},
  {group: 'Primitives', label: 'Text'},
  {group: 'Primitives', label: 'Non-modal dialog'},
  {group: 'Layout', label: 'Grid'},
  {label: 'Design Tokens'},
];

const ComboboxMock: React.FC<{}> = () => {
  const [inputItems, setInputItems] = React.useState(items);
  return (
    <>
      <Combobox
        items={inputItems}
        initialIsOpen
        helpText="This is the help text"
        labelText="Choose a component:"
        onInputValueChange={({inputValue}) => {
          if (inputValue !== undefined) {
            setInputItems(items.filter(item => item.toLowerCase().startsWith(inputValue.toLowerCase())));
          }
        }}
      />
    </>
  );
};

const GroupedMockCombobox: React.FC<{groupLabelTemplate?: ComboboxProps['groupLabelTemplate']}> = ({
  groupLabelTemplate,
}) => {
  return (
    <Combobox
      initialIsOpen
      groupItemsBy="group"
      items={groupedItems}
      labelText="Choose a component:"
      helpText="This is group"
      groupLabelTemplate={groupLabelTemplate}
      optionTemplate={(item: any) => <div>{item.label}</div>}
      itemToString={item => (item && typeof item !== 'string' ? item.label : null)}
    />
  );
};

const ControlledCombobox: React.FC = () => {
  const [value, setValue] = React.useState('');
  const [selectedItem, setSelectedItem] = React.useState({});
  const [inputItems, setInputItems] = React.useState(objectItems);
  const {reset, ...state} = useCombobox({
    items: inputItems,
    itemToString: item => (item ? item.label : null),
    onSelectedItemChange: changes => {
      setSelectedItem(changes.selectedItem);
    },
    onInputValueChange: ({inputValue}) => {
      if (inputValue !== undefined) {
        setInputItems(
          _.filter(objectItems, (item: any) => item.label.toLowerCase().startsWith(inputValue.toLowerCase()))
        );
        setValue(inputValue);
      }
    },
    inputValue: value,
  });
  return (
    <>
      <Combobox
        state={state}
        items={inputItems}
        autocomplete
        labelText="Choose a country:"
        helpText="This is the help text"
        optionTemplate={(item: any) => (
          <div>
            {item.code} | {item.label} | {item.phone}
          </div>
        )}
        insertAfter={
          <Button
            variant="link"
            size="reset"
            onClick={() => {
              reset();
            }}
          >
            <CloseIcon decorative={false} title="Clear" />
          </Button>
        }
      />
      <Box paddingTop="space70">
        Input value state: <span data-testid="input-value-span">{JSON.stringify(value)}</span>
        <br />
        Selected item state: <span data-testid="selected-item-span">{JSON.stringify(selectedItem)}</span>
      </Box>
    </>
  );
};

describe('Combobox ', () => {
  describe('Render', () => {
    it('should render', () => {
      const {asFragment} = render(<ComboboxMock />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render a combobox with aria attributes', () => {
      render(<ComboboxMock />);
      const renderedCombobox = screen.getByRole('combobox');
      expect(renderedCombobox.getAttribute('aria-haspopup')).toEqual('listbox');
      expect(renderedCombobox.getAttribute('aria-owns')).toEqual(screen.getByRole('listbox').id);
      expect(renderedCombobox.getAttribute('aria-expanded')).toEqual('true');
    });

    it('should render a textbox with aria attributes', () => {
      render(<ComboboxMock />);
      const renderedCombobox = screen.getByRole('textbox');
      expect(renderedCombobox.getAttribute('aria-controls')).toEqual(screen.getByRole('listbox').id);
      expect(renderedCombobox.getAttribute('aria-labelledby')).toEqual(document.querySelector('label').id);
      expect(renderedCombobox.getAttribute('aria-describedby')).not.toEqual('');
    });

    it('should render a list with aria attributes', () => {
      render(<ComboboxMock />);
      const renderedCombobox = screen.getByRole('listbox');
      expect(renderedCombobox.getAttribute('aria-labelledby')).toEqual(document.querySelector('label').id);
    });

    it('should render a list with unique option ids', () => {
      render(<ComboboxMock />);
      const renderedOptions = screen.getAllByRole('option');
      const optionIDs = renderedOptions.map(option => option.id);
      const uniqueIDs = _.uniq(optionIDs);
      expect(uniqueIDs.length).toEqual(optionIDs.length);
    });

    it('should render a label for that matches the input id', () => {
      render(<ComboboxMock />);
      const renderedLabel = document.querySelector('label');
      const renderedTextbox = screen.getByRole('textbox');
      expect(renderedLabel.getAttribute('for')).toEqual(renderedTextbox.getAttribute('id'));
    });
  });

  describe('Groups', () => {
    it('should render a group of options', () => {
      render(<GroupedMockCombobox />);
      const renderedGroups = screen.getAllByRole('group');
      // check groups, group label and number of options per group
      expect(renderedGroups[0].getAttribute('aria-label')).toEqual('Components');
      expect(renderedGroups[0].querySelectorAll('[role="option"]').length).toEqual(8);
      expect(renderedGroups[1].getAttribute('aria-label')).toEqual('Primitives');
      expect(renderedGroups[1].querySelectorAll('[role="option"]').length).toEqual(3);
      expect(renderedGroups[2].getAttribute('aria-label')).toEqual('Layout');
      expect(renderedGroups[2].querySelectorAll('[role="option"]').length).toEqual(1);
    });

    it('should render any items not identified as part of the group as ungrouped options', () => {
      render(<GroupedMockCombobox />);
      const renderedListbox = screen.getByRole('listbox');
      const renderedGroups = screen.getAllByRole('group');
      // check we have 3 groups
      expect(renderedGroups.length).toEqual(3);
      // check any options that are not nested in groups
      expect(renderedListbox.querySelectorAll('[role="listbox"] > [role="option"]').length).toEqual(1);
    });

    it('should render a listbox with groups of options that contains no duplicate ids', () => {
      render(<GroupedMockCombobox />);
      const renderedOptions = screen.getAllByRole('option');
      const optionIDs = renderedOptions.map(option => option.id);
      const uniqueIDs = _.uniq(optionIDs);
      expect(uniqueIDs.length).toEqual(optionIDs.length);
    });

    it('should render a custom group label', () => {
      render(<GroupedMockCombobox groupLabelTemplate={groupName => <span>hi {groupName}</span>} />);
      const renderedGroups = screen.getAllByRole('group');
      expect(renderedGroups[0].querySelector('[role="group"] > div[role="presentation"]').textContent).toEqual(
        'hi Components'
      );
      expect(renderedGroups[1].querySelector('[role="group"] > div[role="presentation"]').textContent).toEqual(
        'hi Primitives'
      );
      expect(renderedGroups[2].querySelector('[role="group"] > div[role="presentation"]').textContent).toEqual(
        'hi Layout'
      );
    });
  });

  describe('Controlled Combobox', () => {
    it('should be able to clear a Combobox', () => {
      render(<ControlledCombobox />);
      // open the combobox
      fireEvent.click(screen.getByRole('textbox'));
      // select the first item
      fireEvent.click(screen.getAllByRole('option')[0]);
      // @ts-ignore Property 'value' does not exist on type 'HTMLElement' (I get it, but this is right)
      expect(screen.getByRole('textbox').value).toEqual('Andorra');
      expect(screen.getByTestId('input-value-span').textContent).toEqual('"Andorra"');
      expect(screen.getByTestId('selected-item-span').textContent).toEqual(
        '{"code":"AD","label":"Andorra","phone":"376"}'
      );
      // click the clear button
      fireEvent.click(screen.getByText('Clear'));
      // @ts-ignore Property 'value' does not exist on type 'HTMLElement' (I get it, but this is right)
      expect(screen.getByRole('textbox').value).toEqual('');
      expect(screen.getByTestId('input-value-span').textContent).toEqual('""');
      expect(screen.getByTestId('selected-item-span').textContent).toEqual('null');
    });
  });

  describe('Accessibility', () => {
    it('Should have no accessibility violations', async () => {
      const {container} = render(<ComboboxMock />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
