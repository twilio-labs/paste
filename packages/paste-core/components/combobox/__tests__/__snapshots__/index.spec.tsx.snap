// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox Render should render 1`] = `
<DocumentFragment>
  .emotion-27 {
  box-sizing: border-box;
  position: relative;
}

.emotion-2 {
  box-sizing: border-box;
  border-bottom-width: borderWidth0;
  display: block;
  margin-bottom: space20;
  padding-left: space0;
  padding-right: space0;
  text-transform: none;
}

.emotion-1 {
  box-sizing: border-box;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: size0;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  font-size: fontSize30;
  font-weight: fontWeightSemibold;
  line-height: lineHeight30;
  color: colorText;
  cursor: pointer;
}

.emotion-7 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: colorBackgroundBody;
  box-shadow: shadowBorder;
  border-radius: borderRadius20;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-7:hover {
  box-shadow: shadowBorderPrimaryDark;
}

.emotion-7:focus-within {
  box-shadow: shadowFocus;
}

.emotion-7:active,
.emotion-7[data-active=true] {
  box-shadow: shadowBorderPrimaryDarker;
}

.emotion-6 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  width: 100%;
}

.emotion-3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: borderRadius20;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: fontSize30;
  font-weight: fontWeightNormal;
  line-height: lineHeight20;
  outline: none;
  padding-bottom: space30;
  padding-left: space40;
  padding-right: space40;
  padding-top: space30;
  resize: none;
  width: 100%;
  padding-right: space100;
}

.emotion-3::-webkit-input-placeholder {
  color: colorTextWeak;
  font-style: italic;
}

.emotion-3::-moz-placeholder {
  color: colorTextWeak;
  font-style: italic;
}

.emotion-3:-ms-input-placeholder {
  color: colorTextWeak;
  font-style: italic;
}

.emotion-3::placeholder {
  color: colorTextWeak;
  font-style: italic;
}

.emotion-3:focus::-webkit-input-placeholder {
  color: colorTextWeak;
}

.emotion-3:focus::-moz-placeholder {
  color: colorTextWeak;
}

.emotion-3:focus:-ms-input-placeholder {
  color: colorTextWeak;
}

.emotion-3:focus::placeholder {
  color: colorTextWeak;
}

.emotion-3:disabled {
  color: colorTextWeaker;
  cursor: not-allowed;
}

.emotion-5 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: absolute;
  pointer-events: none;
  right: space30;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  z-index: zIndex10;
}

.emotion-4 {
  box-sizing: border-box;
  line-height: lineHeight0;
  display: block;
  color: colorTextIcon;
  width: sizeIcon30;
  height: sizeIcon30;
}

.emotion-24 {
  box-sizing: border-box;
  background-color: colorBackgroundBody;
  border-radius: borderRadius20;
  border-color: colorBorderLight;
  border-width: borderWidth10;
  border-style: solid;
  box-shadow: shadow;
  list-style-type: none;
  margin: space0;
  max-height: size50;
  overflow-y: auto;
  padding: space0;
  padding-top: space30;
  padding-bottom: space30;
  position: absolute;
  width: 100%;
  z-index: zIndex20;
  margin-top: space20;
}

.emotion-9 {
  box-sizing: border-box;
  background-color: colorBackgroundBody;
  padding: space30;
  cursor: pointer;
  padding-left: space70;
  padding-right: space70;
}

.emotion-8 {
  margin: 0;
  padding: 0;
  color: colorText;
  font-size: fontSize30;
  line-height: lineHeight30;
}

.emotion-26 {
  box-sizing: border-box;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: space30;
  min-width: size0;
}

.emotion-25 {
  margin: 0;
  padding: 0;
  color: colorTextWeak;
  font-size: fontSize30;
  line-height: lineHeight30;
}

<div
    class="emotion-27"
  >
    <label
      class="emotion-2"
      display="block"
      for="downshift-0-input"
      id="downshift-1-label"
    >
      <span
        class="emotion-1"
        display="flex"
      >
        <span
          class="emotion-0"
          color="colorText"
          cursor="pointer"
          font-size="fontSize30"
          font-weight="fontWeightSemibold"
        >
          Choose a component:
        </span>
      </span>
    </label>
    <div
      class="emotion-7"
      cursor="text"
      display="flex"
      width="100%"
    >
      <div
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="downshift-1-menu"
        class="emotion-6"
        display="flex"
        role="combobox"
        width="100%"
      >
        <input
          aria-autocomplete="list"
          aria-controls="downshift-1-menu"
          aria-describedby="1"
          aria-labelledby="downshift-1-label"
          autocomplete="off"
          class="emotion-3"
          id="downshift-0-input"
          tabindex="0"
          type="text"
          value=""
        />
        <div
          class="emotion-5"
          display="inline-flex"
          pointer-events="none"
          transform="translateY(-50%)"
        >
          <span
            class="emotion-4"
            color="colorTextIcon"
            display="block"
          >
            <svg
              aria-hidden="true"
              aria-labelledby="ChevronDownIcon-2"
              height="100%"
              role="img"
              viewBox="0 0 20 20"
              width="100%"
            >
              <path
                d="M6.293 8.293a1 1 0 011.32-.083l.094.083L10 10.585l2.293-2.292a1 1 0 011.32-.083l.094.083a1 1 0 01.083 1.32l-.083.094-3 3a1 1 0 01-1.32.083l-.094-.083-3-3a1 1 0 010-1.414z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
    <ul
      aria-labelledby="downshift-1-label"
      class="emotion-24"
      id="downshift-1-menu"
      role="listbox"
      width="100%"
    >
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-0"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Alert
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-1"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Anchor
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-2"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Button
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-3"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Card
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-4"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Heading
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-5"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          List
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-6"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Modal
        </span>
      </li>
      <li
        aria-selected="false"
        class="emotion-9"
        cursor="pointer"
        id="downshift-1-item-7"
        role="option"
      >
        <span
          class="emotion-8"
          color="colorText"
          font-size="fontSize30"
        >
          Paragraph
        </span>
      </li>
    </ul>
    <div
      class="emotion-26"
      display="flex"
    >
      <div
        class="emotion-25"
        color="colorTextWeak"
        font-size="fontSize30"
        id="1"
      >
        This is the help text
      </div>
    </div>
  </div>
</DocumentFragment>
`;
